### PR TITLE
feat: pushdown limit to single window

### PR DESCRIPTION
**Pull request by ariesdevil**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

when single window with frame type `rows` in select list with limit, we pushdown limit to window order by sort.

- Fixes #[Link the issue here]

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14460)

### CODE REVIEW

File: tests/sqllogictests/suites/mode/standalone/explain/window.test [hunk 1]
```diff
@@ -285,5 +285,63 @@ Filter
...
+create table t(a int)
+
+statement ok
+insert into t values(1), (2), (3), (3)
+
+# range frame
+query T
+explain pipeline select a, sum(a) over (partition by a order by a desc) from t limit 3
+----
+CompoundBlockOperator(Project) × 1 processor
```
Comment (sundy-li) [inline tests/sqllogictests/suites/mode/standalone/explain/window.test:302]: I don't know if it's pushed or not from this pipeline.

File: src/query/sql/src/executor/physical_plans/physical_window.rs [hunk 2]
```diff
@@ -55,6 +55,7 @@ pub struct Window {
     pub partition_by: Vec<IndexType>,
     pub order_by: Vec<SortDesc>,
     pub window_frame: WindowFuncFrame,
+    pub limit: Option<usize>,
```
Comment (Dousir9) [inline src/query/sql/src/executor/physical_plans/physical_window.rs:58]: We can add `limit:xxx` to `window_to_format_tree` in `src/query/sql/src/executor/format.rs` to display limit in explain.

File: tests/sqllogictests/suites/mode/standalone/explain/window.test (hunk 1 above)
Comment (ariesdevil) [reply tests/sqllogictests/suites/mode/standalone/explain/window.test:302]: Yep, for range frame we cannot push limit to window sorts, the following rows frame will change `TransformSortMerge` => `TransformSortMergeLimit`

Comment (ariesdevil) [reply tests/sqllogictests/suites/mode/standalone/explain/window.test:302]: I'll add the ranking functions cause they are frame insensitive.

Comment (xudong963) [reply tests/sqllogictests/suites/mode/standalone/explain/window.test:302]: why not use explain select but use explain pipeline?

File: src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_limit_window.rs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        Ok(())
+    }
+
+    fn patterns(&self) -> &Vec<SExpr> {
+        &self.patterns
+    }
+}
+
+fn should_apply(child: &SExpr, window: &LogicalWindow) -> Result<bool> {
+    let child_window_exists = child_has_window(child)?;
```
[the remark this answers is not in the document]
Comment (ariesdevil) [reply src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_limit_window.rs:106]: For simply, if we have multiple windows, they may have different partition by and order by，then we should check they are same or they both can apply limit

File: tests/sqllogictests/suites/mode/standalone/explain/window.test (hunk 1 above)
Comment (ariesdevil) [reply tests/sqllogictests/suites/mode/standalone/explain/window.test:302]: the explain are same，we just reduce the sort with limit for window input，but still need sort

**Review by sundy-li — APPROVED** [review]

**Comment by BohuTANG** [comment]
Conflicting files:
src/query/service/src/pipelines/builders/builder_window.rs
src/query/sql/src/executor/format.rs

**Pull request merged**